File: medium/3720-lexicographically-smallest-permutation-greater-than-target/solution.py
class Solution:
    def smallestPermutation(self, s: str, target: str) -> str:
        """
        3720. Lexicographically Smallest Permutation Greater Than Target
        Time: O(26 * n)
        Space: O(n)
        """
        n = len(s)
        counts = [0] * 26
        for ch in s:
            counts[ord(ch) - 97] += 1

        # Longest prefix of target that s's letters can spell.
        matched = 0
        while matched < n and counts[ord(target[matched]) - 97] > 0:
            counts[ord(target[matched]) - 97] -= 1
            matched += 1

        # Prefer the longest shared prefix: deviating later gives a smaller result.
        for p in range(matched, -1, -1):
            # Roll characters back into the pool until counts match a prefix of length p.
            while matched > p:
                matched -= 1
                counts[ord(target[matched]) - 97] += 1

            if p >= n:
                continue  # a full match is equal, not strictly greater

            # Smallest available letter strictly greater than target[p].
            for c in range(ord(target[p]) - 97 + 1, 26):
                if counts[c] > 0:
                    counts[c] -= 1
                    rest = "".join(chr(97 + j) * counts[j] for j in range(26))
                    return target[:p] + chr(97 + c) + rest

        return ""

File: medium/3720-lexicographically-smallest-permutation-greater-than-target/solution.java
/**
 * 3720. Lexicographically Smallest Permutation Greater Than Target
 * Time: O(26 * n)
 * Space: O(n)
 */
class Solution {
    public String smallestPermutation(String s, String target) {
        int n = s.length();
        int[] counts = new int[26];
        for (int i = 0; i < n; i++) {
            counts[s.charAt(i) - 'a']++;
        }

        // Longest prefix of target that s's letters can spell.
        int matched = 0;
        while (matched < n && counts[target.charAt(matched) - 'a'] > 0) {
            counts[target.charAt(matched) - 'a']--;
            matched++;
        }

        // Prefer the longest shared prefix: deviating later gives a smaller result.
        for (int p = matched; p >= 0; p--) {
            // Roll characters back into the pool until counts match a prefix of length p.
            while (matched > p) {
                matched--;
                counts[target.charAt(matched) - 'a']++;
            }

            if (p >= n) {
                continue; // a full match is equal, not strictly greater
            }

            // Smallest available letter strictly greater than target[p].
            for (int c = target.charAt(p) - 'a' + 1; c < 26; c++) {
                if (counts[c] > 0) {
                    counts[c]--;
                    StringBuilder sb = new StringBuilder();
                    sb.append(target, 0, p);
                    sb.append((char) ('a' + c));
                    for (int j = 0; j < 26; j++) {
                        for (int t = 0; t < counts[j]; t++) {
                            sb.append((char) ('a' + j));
                        }
                    }
                    return sb.toString();
                }
            }
        }

        return "";
    }
}
